modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/idempotency/NuxeoIdempotentFilter.java

```diff
@@ -0,0 +1,168 @@
+        HttpServletResponse response = (HttpServletResponse) servletResponse;
+        String method = request.getMethod();
+        if (isIdempotentMethod(method)) {
+            log.debug("No idempotent processing done: method is already idempotent: " + method);
+            chain.doFilter(request, response);
+            return;
+        }
+        String key = request.getHeader(NuxeoIdempotentFilter.HEADER_KEY);
+        if (key == null) {
+            log.debug("No idempotent processing done: no Idempotent-Key present");
+            chain.doFilter(request, response);
+            return;
+        }
+        log.debug("Idempotent request key: " + key);
+        KeyValueService kvs = Framework.getService(KeyValueService.class);
+        if (kvs == null) {
+            log.debug("KeyValueService not present");
+            chain.doFilter(request, response);
+            return;
+        }
+        KeyValueStore store = kvs.getKeyValueStore(getStoreName());
+        String status = store.getString(key + NuxeoIdempotentFilter.INFO_SUFFIX);
+        if (status == null) {
+            log.debug("Handle new request for key: " + key);
+            long ttl = getTTLSeconds();
+            store.put(key + NuxeoIdempotentFilter.INFO_SUFFIX, NuxeoIdempotentFilter.INPROGRESS_MARKER, ttl);
+            try {
+                CopyingResponseWrapper wrapper = new CopyingResponseWrapper(DEFERRED_OUTPUT_STREAM_THRESHOLD, response);
+                chain.doFilter(request, wrapper);
+                store.put(key, wrapper.getCaptureAsBytes(), ttl);
+                store.put(key + NuxeoIdempotentFilter.INFO_SUFFIX, NuxeoIdempotentResponse.write(wrapper), ttl);
+                log.debug("Stored response for key: " + key);
+            } catch (IOException | ServletException e) {
+                response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+                throw e;
+            } finally {
+                if (response.getStatus() >= HttpServletResponse.SC_BAD_REQUEST) {
+                    // error request: cleanup store
+                    store.put(key, (String) null);
+                    store.put(key + NuxeoIdempotentFilter.INFO_SUFFIX, (String) null);
+                    log.debug("Cleanup store: error for key: " + key);
+                }
+            }
+        } else if (NuxeoIdempotentFilter.INPROGRESS_MARKER.equals(status)) {
+            // request already in progress -> conflict
+            // Don't call response.sendError, because it commits the response
+            // which prevents NuxeoExceptionFilter from returning a custom error page.
+            response.setStatus(HttpServletResponse.SC_CONFLICT);
+            log.debug("Conflict response for key: " + key);
+        } else {
+            // request already done: return stored result
+            response.getOutputStream().write(store.get(key));
+            NuxeoIdempotentResponse.restore(response, store.get(key + NuxeoIdempotentFilter.INFO_SUFFIX));
```

Comment:
You should do this first, and if there's an `IOException` (due to errors parsing the JSON) log it but abandon everything and process normally the chain.